File: app/src/main/java/com/developer/valyutaapp/ui/valute/ValuteActivity.java
package com.developer.valyutaapp.ui.valute;

import android.content.Context;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.developer.valyutaapp.R;
import com.developer.valyutaapp.adapter.DialogAdapter;
import com.developer.valyutaapp.model.Valute;
import com.developer.valyutaapp.utils.ImageResource;
import com.developer.valyutaapp.utils.Utils;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class ValuteActivity extends AppCompatActivity implements ValuteViewInterface, DialogAdapter.ClickListener{

    AlertDialog.Builder alertdialog;
    AlertDialog dialog;
    ArrayList<Valute> valutes;
    private long valuteId = 0;
    DialogAdapter adapter;

    @BindView(R.id.iconValute1) ImageView iconValute1;
    @BindView(R.id.iconValute2) ImageView iconValute2;
    @BindView(R.id.clear) ImageView clear;
    @BindView(R.id.change) ImageView change;
    @BindView(R.id.edit1) EditText value1;
    @BindView(R.id.edit2) EditText value2;

    @BindView(R.id.name1) TextView code1;
    @BindView(R.id.name2) TextView code2;
    @Nullable
    @BindView(R.id.progressBar) ProgressBar progressBar;

    public static double nominal = 0;
    public static double value = 0;

    //LineChart chart;

    private ValutePresenter valutePresenter;

    private String TAG = "ValuteActivity";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_valute);
        ButterKnife.bind(this);
        Bundle extras = getIntent().getExtras();
        if (extras != null) {
            valuteId = extras.getInt("id");
        }

        valutes = new ArrayList<>();

        setupMVP();
        getValuteById();
        getValuteList();

        clear.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                value1.setText("");
                value2.setText("");
            }
        });
        change.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (value1.isEnabled() && hasWindowFocus()){
                    value2.requestFocus();
                }
                if (!value1.isEnabled() && !hasWindowFocus()){
                    value1.requestFocus();
                }
            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();

        if (value1.isEnabled()) {
            value1.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                }

                @Override
                public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                }

                @Override
                public void afterTextChanged(Editable editable) {
                    if (editable.length() == 0) {
                        return;
                    } else if (code2.getText().equals("TJS")) {
                        double sum = 0;
                        nominal = Double.parseDouble(String.valueOf(value1.getText()));
                        sum = Utils.mathNominal(nominal, value);
                        Log.d("sum", nominal + "*" + value + " = " + sum);
                        value2.setText(String.valueOf(sum));
                    }
                }
            });
        }else {
            value2.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                }

                @Override
                public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                }

                @Override
                public void afterTextChanged(Editable editable) {
                    if (editable.length() == 0) {
                        return;
                    } else if (code2.getText().equals("TJS")) {
                        double sum = 0, valueDynamic;
                        valueDynamic = Double.parseDouble(String.valueOf(value2.getText()));
                        sum = Utils.mathValue(nominal, value, valueDynamic);
                        Log.d("sum", nominal + "*" + value + " = " + sum);
                        value1.setText(String.valueOf(sum));
                    }
                }
            });
        }
    }

    private void setupMVP() {
        valutePresenter = new ValutePresenter(this.getApplicationContext(), this);
    }

    private void getValuteById() {
        valutePresenter.getValuteById((int) valuteId);
    }

    private void getValuteList() {
        valutePresenter.getValutes();
    }

    @Override
    public void showToast(String str) {
        Toast.makeText(ValuteActivity.this,str,Toast.LENGTH_LONG).show();
    }

    @Override
    public void showProgressBar() {
        progressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideProgressBar() {
        progressBar.setVisibility(View.GONE);
    }

    @Override
    public void displayValuteWithId(Valute valute) {
        Bitmap bt = ImageResource.getImageRes(this, String.valueOf(valute.getCharCode()));
        iconValute1.setImageBitmap(bt);
        value = Double.parseDouble(valute.getValue());
        value1.setText(String.valueOf(valute.getNominal()));
        value2.setText(String.valueOf(valute.getValue()));
        code1.setText(String.valueOf(valute.getCharCode()));
        code2.setText("TJS");
        iconValute2.setImageResource(R.drawable.tajikistan);
    }

    @Override
    public void displayValutes(List<Valute> valute) {
        valutes.addAll(valute);
    }

    @Override
    public void displayError(String e) {
        showToast(e);
    }

    public void dialogValutes(View v){
        alertdialog = new AlertDialog.Builder(this);
        alertdialog.setTitle("Все валюты");
        LayoutInflater inflater = (LayoutInflater) this.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        View row = inflater.inflate(R.layout.row_item, null);
        ListView listView = (ListView) row.findViewById(R.id.list_dialog);
        adapter = new DialogAdapter(this, valutes);
        adapter.setClickListener(this);
        listView.setAdapter(adapter);
        alertdialog.setView(row);
        dialog = alertdialog.create();
        dialog.show();
    }

    @Override
    public void itemClicked(Valute item, int position) {
        valuteId = item.getId();
        dialog.hide();
        getValuteById();
    }
}

File: app/src/main/java/com/developer/valyutaapp/ui/valute/ValuteActivity.kt
package com.developer.valyutaapp.ui.valute

import android.annotation.SuppressLint
import androidx.appcompat.app.AppCompatActivity
import com.developer.valyutaapp.domain.entities.Valute
import android.os.Bundle
import android.text.TextWatcher
import android.text.Editable
import android.util.Log
import com.developer.valyutaapp.utils.ImageResource
import android.view.LayoutInflater
import android.view.View
import android.widget.*
import androidx.appcompat.app.AlertDialog
import by.kirich1409.viewbindingdelegate.viewBinding
import com.developer.valyutaapp.R
import com.developer.valyutaapp.utils.Utils
import com.developer.valyutaapp.databinding.ActivityValuteBinding
import java.util.ArrayList

class ValuteActivity : AppCompatActivity(R.layout.activity_valute), ValuteViewInterface {

    private val viewBinding by viewBinding(ActivityValuteBinding::bind, R.id.container)

    var dialog: AlertDialog? = null
    private var valuteId: Long = 0

    private var valutes: MutableList<Valute> = ArrayList()
    private val adapter by lazy {  }

    //LineChart chart;
    //private var valutePresenter: ValutePresenter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val extras = intent.extras
        if (extras != null) {
            valuteId = extras.getInt("id").toLong()
        }
        valutes = ArrayList()
        setupMVP()
        valuteById
        valuteList
        viewBinding.clear.setOnClickListener {
            viewBinding.edit1.setText("")
            viewBinding.edit2.setText("")
        }
        viewBinding.change.setOnClickListener {
            if (viewBinding.edit1.isEnabled && hasWindowFocus()) {
                viewBinding.edit2.requestFocus()
            }
            if (!viewBinding.edit1.isEnabled && !hasWindowFocus()) {
                viewBinding.edit1.requestFocus()
            }
        }
    }

    override fun onResume() {
        super.onResume()
        if (viewBinding.edit1.isEnabled) {
            viewBinding.edit1.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(
                    charSequence: CharSequence,
                    i: Int,
                    i1: Int,
                    i2: Int
                ) {
                }

                override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
                override fun afterTextChanged(editable: Editable) {
                    if (editable.isEmpty()) {
                        return
                    } else if (viewBinding.name1.text == "TJS") {
                        nominal = viewBinding.name1.text.toString().toDouble()
                        val sum = Utils.mathNominal(nominal, value)
                        Log.d("sum", "$nominal*$value = $sum")
                        viewBinding.edit2.setText(sum.toString())
                    }
                }
            })
        } else {
            viewBinding.edit2.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(
                    charSequence: CharSequence,
                    i: Int,
                    i1: Int,
                    i2: Int
                ) {
                }

                override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
                override fun afterTextChanged(editable: Editable) {
                    if (editable.isEmpty()) {
                        return
                    } else if (viewBinding.name2.text == "TJS") {
                        val valueDynamic: Double = viewBinding.name2.text.toString().toDouble()
                        val sum = Utils.mathValue(nominal, value, valueDynamic)
                        Log.d("sum", "$nominal*$value = $sum")
                        viewBinding.edit1.setText(sum.toString())
                    }
                }
            })
        }
    }

    private fun setupMVP() {
        //valutePresenter = ValutePresenter(this.applicationContext, this)
    }

    private val valuteById: Unit
        get() {
           // valutePresenter!!.getValuteById(valuteId.toInt())
        }
    private val valuteList: Unit
        get() {
            //valutePresenter!!.valutes()
        }

    override fun showToast(s: String) {
        Toast.makeText(this@ValuteActivity, s, Toast.LENGTH_LONG).show()
    }

    override fun showProgressBar() {
        viewBinding.progressBar.visibility = View.VISIBLE
    }

    override fun hideProgressBar() {
        viewBinding.progressBar.visibility = View.GONE
    }

    @SuppressLint("SetTextI18n")
    override fun displayValuteWithId(valute: Valute) {
        val bt = ImageResource.getImageRes(this, valute.charCode)
        viewBinding.iconValute1.setImageBitmap(bt)
        value = valute.value.toDouble()
        viewBinding.edit1.setText(java.lang.String.valueOf(valute.nominal))
        viewBinding.edit2.setText(valute.value)
        viewBinding.name1.text = valute.charCode
        viewBinding.name2.text = "TJS"
        viewBinding.iconValute2.setImageResource(R.drawable.tajikistan)
    }

    override fun displayValutes(valute: List<Valute>) {
        valutes.addAll(valute)
    }

    override fun displayError(s: String) {
        showToast(s)
    }

    fun dialogValutes(v: View?) {
       val builder = AlertDialog.Builder(this)
        builder.setTitle("Все валюты")
        val inflater = this.getSystemService(LAYOUT_INFLATER_SERVICE) as LayoutInflater
        val row = inflater.inflate(R.layout.row_item, null)
        val listView = row.findViewById<View>(R.id.list_dialog) as ListView
        //adapter = DialogAdapter(this, valutes!!)
       // listView.adapter = adapter
        builder.setView(row)
        dialog = builder.create()
        dialog!!.show()
    }

     fun itemClicked(item: Valute?, position: Int) {
        valuteId = item!!.id.toLong()
        dialog!!.hide()
        valuteById
    }

    companion object {
        var nominal = 0.0
        var value = 0.0
    }
}